indigo/indigo-extras/src/main/scala/indigoextras/ui/RadioButton.scala

```diff
@@ -0,0 +1,206 @@
+    this.copy(onUnselected = action)
+
+  /**
+    * In constructing this radio button, set the events to fire on an option gaining selection.
+    *
+    *  @param action A mapping from an option index to the list of events to fire
+    * @return The updated radio button construct
+    */
+  def withSelectedAction(action: Int => List[GlobalEvent]): RadioButton =
+    this.copy(onSelected = action)
+
+  /**
+    * In constructing this radio button, set the events to fire on an option being hovered over.
+    *
+    * @param action A mapping from an option index to the list of events to fire.
+    * @return The updated radio button construct
+    */
+  def withHoverOverAction(action: Int => List[GlobalEvent]): RadioButton =
+    this.copy(onHoverOver = action)
+
+  /**
+    * In constructing this radio button, set the events to fire on an option losing hover.
+    *
+    * @param action A mapping from an option index to the list of events to fire
+    * @return The updated radio button construct
+    */
+  def withHoverOutAction(action: Int => List[GlobalEvent]): RadioButton =
+    this.copy(onHoverOut = action)
+}
+
+object RadioButton {
+
+  /**
+    * Construct a radio button without yet specifying the events fired on selection, hover, and so on,
+    * with the bounds of each option being specified by the rectangle they occupy.
+    *
+    * @param buttonAssets The button assets to use to present each option button
+    * @param bounds The rectangular bounds of each option button by index
+    * @param depth The display depth at which to present the radio button
+    * @param selected Which option is initially selected, if any
+    * @return The constructed radio button
+    */
+  def apply(buttonAssets: ButtonAssets, bounds: List[Rectangle], depth: Depth, selected: Option[Int]): RadioButton =
+    RadioButton(
+      buttonAssets = buttonAssets,
+      bounds = bounds,
+      depth = depth,
+      selected = selected,
+      over = None,
+      onUnselected = _ => Nil,
+      onSelected = _ => Nil,
+      onHoverOut = _ => Nil,
+      onHoverOver = _ => Nil
+    )
+
+  /**
+    * Construct a radio button without yet specifying the events fired on selection, hover, and so on,
+    * with the bounds of each option being specified by the top-left position of each plus the
+    * width and height to use for them all.
+    *
+    * @param buttonAssets The button assets to use to present each option button
+    * @param positions The top-left position of each option button by index
+    * @param width The width of an option button, identical for all
+    * @param height The height of an option button, identical for all
+    * @param depth The display depth at which to present the radio button
+    * @param selected Which option is initially selected, if any
+    * @return The constructed radio button
+    */
+  def apply(
+      buttonAssets: ButtonAssets,
+      positions: List[Point],
```

Comment:
I mean that the width and the heights are encoded into the `ButtonAssets` aren't they?

Reply:
Similar to my comment on the example, I was aiming to be as consistent as possible between Button and RadioButton to give a uniform interface. The same comment that width and height are in ButtonAssets seems to apply to Button, which requires Rectangle bounds on construction. Happy to do either - change to using Points everywhere or retain consistency with Button - as you prefer.

Reply:
I appreciate the thoughtful rationale, but I wouldn't assume that just because I did a thing, it was a good thing to do. The button example is probably one of the oldest ones and in need of some love! :-)

I would prefer that your types modelled your domain as strictly as you can (stopping just short of doing something potentially off putting to beginners), rather than worry too much about consistency. Well typed first, consistent second.

So I'd go with Point in this case.

Reply:
I'm trying to implement this change, but if I construct a radio button using Points I still need to know the bounds of radio buttons in order to detect whether they are pressed/hovered over. You say above that these are encoded in ButtonAssets, which makes sense, but I'm not clear how to extract them. A button graphic within ButtonAssets, i.e. a SceneGraphNodePrimitive, can return its bounds, but requires a BoundaryLocator, and I'm not sure where I'd derive one from?

Reply:
Taking the last point first: The boundary locator is in the frame context. A specific note on bounds checks with graphics: They're very cheap as there is no actual calculation to do, the BoundaryLocator requirement is just to satisfy the interface. You can also call `lazyBounds`.

However:
1. I hadn't realised the problem in all honesty - you make an excellent point.
2. In the button code I've literally done the same thing as you have, and used bounds. I must have had a similar problem (or not even realised I'd done that).

Therefore, please accept my apologies and carry on.

Reply:
:-) No problem. In the most recent commit, I've changed to giving the Points for the positions plus a (Int) width and height on construction rather than Rectangle bounds, as that seems easiest/most minimal for the client. If you later make width and height accessible from ButtonAssets, then these width and height parameters could be removed from the constructor. Hope that makes sense!

Reply:
There is a secret advantage to forcing the game developer to specify the size, now I come to think of it.

It's trivial for the developer to use `buttonAssets.up.lazyBounds.size.width/height` to set the dimensions (should that be `size: Point` instead of `width` and `height`? Up to you). So it's not a big burden.

The advantage is that the bounds are not constrained by the graphics. You could have a large graphic with only a small part of it acting as the radio button, or a small graphic with a large hit area improving usability.